frontend/packages/console-dynamic-plugin-sdk/src/extensions/user-preferences.ts

```diff
@@ -0,0 +1,84 @@
+import * as React from 'react';
+import { Extension, ExtensionDeclaration, CodeRef } from '../types';
+
+export enum UserPreferenceFieldType {
+  dropdown = 'dropdown',
+  checkbox = 'checkbox',
+  custom = 'custom',
+}
+
+export type UserPreferenceDropdownField = {
+  type: UserPreferenceFieldType.dropdown;
+  userSettingsKey: string;
+  defaultValue?: string;
+  options: {
+    value: string;
+    label: string;
+  }[];
+};
+
+export type UserPreferenceCheckboxFieldValue = string | number | boolean;
+
+export type UserPreferenceCheckboxField = {
+  type: UserPreferenceFieldType.checkbox;
+  userSettingsKey: string;
+  label: string;
+  trueValue: UserPreferenceCheckboxFieldValue;
+  falseValue: UserPreferenceCheckboxFieldValue;
+};
+
+export type UserPreferenceCustomField = {
+  type: UserPreferenceFieldType.custom;
+  component: CodeRef<React.ComponentType>;
+  // TODO: update type from any to json compatible types
+  props?: { [key: string]: any };
```

Comment:
What about `JSONSchema7` from `json-schema`?

Reply:
If I'm not getting it wrong, I think you mean `JSONSchema7Type`, and i think it will be a great fit for the prop type.